HeavyFlavorAnalysis/RecoDecay/src/BPHDecayMomentum.cc

```diff
@@ -0,0 +1,272 @@
+  return nComp;
+}
+
+
+const vector<const reco::Candidate*>& BPHDecayMomentum::daughters() const {
+  return dList;
+}
+
+
+const vector<const reco::Candidate*>& BPHDecayMomentum::daughFull() const {
+  // compute total momentum to update the full list of decay products
+  computeMomentum();
+  return dFull;
+}
+
+
+const reco::Candidate* BPHDecayMomentum::originalReco(
+                       const reco::Candidate* daug ) const {
+  // return the original particle for a given clone
+  // return null pointer if not found
+  map<const reco::Candidate*,
+      const reco::Candidate*>::const_iterator iter = clonesMap.find( daug );
+  return ( iter != clonesMap.end() ? iter->second : 0 );
+}
+
+
+const vector<BPHRecoConstCandPtr>& BPHDecayMomentum::daughComp() const {
+  // return the list of previously reconstructed particles
+  return cList;
+}
+
+
+const reco::Candidate* BPHDecayMomentum::getDaug(
+                       const string& name ) const {
+  // return a simple particle from the name
+  // return null pointer if not found
+  string::size_type pos = name.find( "/" );
+  if ( pos != string::npos ) return getComp( name.substr( 0, pos ) )
+                                  ->getDaug( name.substr( pos + 1 ) );
+  map<const string,
+      const reco::Candidate*>::const_iterator iter = dMap.find( name );
+  return ( iter != dMap.end() ? iter->second : 0 );
+}
+
+
+BPHRecoConstCandPtr BPHDecayMomentum::getComp( const string& name ) const {
+  // return a previously reconstructed particle from the name
+  // return null pointer if not found
+  map<const string,
+      BPHRecoConstCandPtr>::const_iterator iter = cMap.find( name );
+  return ( iter != cMap.end() ? iter->second : 0 );
+}
+
+
+const vector<BPHDecayMomentum::Component>&
+             BPHDecayMomentum::componentList() const {
+  // return an object filled in the constructor
+  // to be used in the creation of other bases of BPHRecoCandidate
+  return compList;
+}
+
+
+void BPHDecayMomentum::setNotUpdated() const {
+  updated = false;
+  return;
+}
+
+
+void BPHDecayMomentum::clonesList( const map<string,Component>& daugMap ) {
+  int n = daugMap.size();
+  dList.resize( n );
+  nList.resize( n );
+  // reset and fill a list
+  // to be used in the creation of other bases of BPHRecoCandidate
+  compList.clear();
+  compList.reserve( n );
+  // loop over daughters
+  int i = 0;
+  double mass;
+  reco::Candidate* dnew;
+  map<string,Component>::const_iterator iter = daugMap.begin();
+  map<string,Component>::const_iterator iend = daugMap.end();
+  while ( iter != iend ) {
+    const pair<string,Component>& entry = *iter++;
+    const Component& comp = entry.second;
+    const reco::Candidate* cand = comp.cand;
+    // store component for usage
+    // in the creation of other bases of BPHRecoCandidate
+    compList.push_back( comp );
+    // clone particle and store it with its name
+    dList[i] = dnew = cand->clone();
+    dMap[nList[i++] = entry.first] = dnew;
```

Comment:
if this is not a typo ```aMap[a = b] = c```, a more verbose implementation with a temporary corresponding to the result of ``` a = b``` would be more readable.